toplevel/toploop.ml

```diff
@@ -184,6 +184,54 @@ let load_ocamlinit ppf =
 
 exception PPerror
 
+(* Test whether a lexbuf contains only blank characters, including at least
+   one linefeed. See newline and blank in lexer.mll. *)
+let is_blank_with_linefeed lb =
+  let rec loop i lf_seen =
+    if i >= lb.Lexing.lex_buffer_len then lf_seen else
+      match Bytes.get lb.Lexing.lex_buffer i with
+      | '\009' | '\012' | '\013' | ' ' -> loop (i+1) lf_seen
+      | '\010' -> loop (i+1) true
+      | _ -> false
+  in
+  loop lb.Lexing.lex_curr_pos false
+
+(* Read and parse toplevel phrases, stop when a complete phrase has been
+   parsed and the lexbuf contains and end of line with optional whitespace. *)
+let rec get_phrases ppf lb phrs =
+  match !parse_toplevel_phrase lb with
+  | phr ->
+    if is_blank_with_linefeed lb then
+      List.rev (phr :: phrs)
+    else
+      get_phrases ppf lb (phr :: phrs)
+  | exception Exit -> raise PPerror
+  | exception e -> Location.report_exception ppf e; []
+
+(* Type, compile and execute a phrase. *)
+let process_phrase ppf snap phr =
+  snap := Btype.snapshot ();
+  Warnings.reset_fatal ();
+  let phr = preprocess_phrase ppf phr in
+  Env.reset_cache_toplevel ();
+  ignore(execute_phrase true ppf phr)
+
+(* Type, compile and execute a list of phrases, setting the report printer
+   to batch mode for all but the first one.
+   We have to use batch mode for reporting for two reasons:
+   1. we can't underline several parts of the input line(s) in place
+   2. the execution of the first phrase may mess up the line count so we
+      can't move the cursor back to the correct line
+ *)
+let process_phrases ppf snap phrs =
+  match phrs with
+  | [] -> ()
+  | phr :: rest ->
+    process_phrase ppf snap phr;
+    Misc.protect_refs
+      Location.[R (report_printer, fun () -> batch_mode_printer)]
+      (fun () -> List.iter (process_phrase ppf snap) rest)
```

Comment:
Nice. (We could use `if rest <> [] then Misc.protect_refs ...` but there is no strong need.)

Reply:
I've followed your suggestion.